zeppelin-server/src/main/java/org/apache/zeppelin/socket/NotebookServer.java

```diff
@@ -2021,4 +2021,46 @@ private void broadcastToWatchers(String noteId, String subject, Message message)
       }
     }
   }
+
+  @Override
+  public void onParaInfosReceived(String noteId, String paragraphId,
+      String interpreterSettingId, Map<String, String> metaInfos) {
+    Note note = notebook().getNote(noteId);
+    if (note != null) {
+      Paragraph paragraph = note.getParagraph(paragraphId);
+      if (paragraph != null) {
+        InterpreterSetting setting = notebook().getInterpreterFactory()
+                                               .get(interpreterSettingId);
+        setting.addNoteToPara(noteId, paragraphId);
+        metaInfos.remove("noteId");
+        metaInfos.remove("paraId");
+        String label = metaInfos.get("label");
+        metaInfos.remove("label");
+        paragraph.updateRuntimeInfos(label, metaInfos, setting.getGroup(), setting.getId());
+        broadcast(note.getId(), new Message(OP.PARAGRAPH).put("paragraph", paragraph));
```

Comment:
I think the paragraph need not be completely broadcasted here. A new message can be created to updated only the runtime infos. Any thoughts on this? @Leemoonsoo @zjffdu